File: server/telegram_bot.py
#!/usr/bin/env python3
"""Telegram bot for Smart Home login verification.

Environment variables:
  TELEGRAM_BOT_TOKEN   — token from @BotFather
  SMART_HOME_URL       — base URL of the app (e.g. https://your-domain.com)
  TELEGRAM_BOT_SECRET  — must match TELEGRAM_BOT_SECRET in app env (optional)
"""

import os
import sys
import json
import time
import urllib.request
import urllib.error
import urllib.parse

BOT_TOKEN = os.environ.get("TELEGRAM_BOT_TOKEN", "")
APP_URL = os.environ.get("SMART_HOME_URL", "http://localhost:3000").rstrip("/")
BOT_SECRET = os.environ.get("TELEGRAM_BOT_SECRET", "")

if not BOT_TOKEN:
    print("TELEGRAM_BOT_TOKEN is not set")
    sys.exit(1)

API = f"https://api.telegram.org/bot{BOT_TOKEN}"


def send_message(chat_id, text):
    data = json.dumps({"chat_id": chat_id, "text": text, "parse_mode": "HTML"}).encode()
    req = urllib.request.Request(
        f"{API}/sendMessage",
        data=data,
        headers={"Content-Type": "application/json"},
    )
    try:
        urllib.request.urlopen(req, timeout=10)
    except Exception as e:
        print(f"[Bot] Send error: {e}")


def verify_code(code, user):
    url = f"{APP_URL}/api/auth/telegram/callback"
    payload = json.dumps({
        "code": code,
        "telegram_user": user.get("first_name", "") + (" " + user["last_name"] if user.get("last_name") else ""),
        "telegram_id": user["id"],
    }).encode()

    headers = {"Content-Type": "application/json"}
    if BOT_SECRET:
        headers["Authorization"] = f"Bearer {BOT_SECRET}"

    req = urllib.request.Request(url, data=payload, headers=headers)
    try:
        resp = urllib.request.urlopen(req, timeout=10)
        result = json.loads(resp.read())
        return result.get("success", False)
    except Exception as e:
        print(f"[Bot] Callback error: {e}")
        return False


def handle_message(msg):
    chat_id = msg["chat"]["id"]
    text = (msg.get("text") or "").strip()
    user = msg.get("from", {})

    if text == "/start":
        send_message(chat_id, "Привет! Отправь мне код с экрана входа в Умный дом.")
        return

    code = text.upper().replace(" ", "")
    if len(code) != 6 or not code.isalnum():
        send_message(chat_id, "Отправь 6-значный код с экрана входа.")
        return

    if verify_code(code, user):
        send_message(chat_id, "Вход выполнен. Можешь вернуться в браузер.")
    else:
        send_message(chat_id, "Код неверный или истёк. Попробуй ещё раз.")


def poll():
    offset = 0
    print("[TelegramBot] Starting...")
    print(f"[TelegramBot] App URL: {APP_URL}")

    while True:
        try:
            url = f"{API}/getUpdates?offset={offset}&timeout=30"
            resp = urllib.request.urlopen(url, timeout=35)
            data = json.loads(resp.read())

            if data.get("ok") and data.get("result"):
                for update in data["result"]:
                    offset = update["update_id"] + 1
                    if "message" in update:
                        handle_message(update["message"])
        except KeyboardInterrupt:
            print("\n[TelegramBot] Stopped.")
            break
        except Exception as e:
            print(f"[Bot] Poll error: {e}")
            time.sleep(5)


if __name__ == "__main__":
    poll()

File: server/telegram-bot.js
// Telegram bot for Smart Home login verification
// Uses only built-in Node.js fetch (Node 18+)
//
// Environment variables:
//   TELEGRAM_BOT_TOKEN  — token from @BotFather
//   SMART_HOME_URL      — base URL of the app (e.g. https://your-domain.com)
//   TELEGRAM_BOT_SECRET — must match TELEGRAM_BOT_SECRET in .env (optional)

const BOT_TOKEN = process.env.TELEGRAM_BOT_TOKEN;
const APP_URL = (process.env.SMART_HOME_URL || "http://localhost:3000").replace(/\/$/, "");
const BOT_SECRET = process.env.TELEGRAM_BOT_SECRET || "";

if (!BOT_TOKEN) {
  console.error("TELEGRAM_BOT_TOKEN is not set");
  process.exit(1);
}

const API = `https://api.telegram.org/bot${BOT_TOKEN}`;
let offset = 0;

async function sendMessage(chatId, text, opts = {}) {
  await fetch(`${API}/sendMessage`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ chat_id: chatId, text, parse_mode: "HTML", ...opts }),
  });
}

async function handleMessage(msg) {
  const chatId = msg.chat.id;
  const text = (msg.text || "").trim();
  const user = msg.from;

  if (text === "/start") {
    await sendMessage(
      chatId,
      "Привет! Отправь мне код с экрана входа в Умный дом."
    );
    return;
  }

  // Expect a 6-character code
  const code = text.toUpperCase().replace(/\s/g, "");
  if (!/^[A-Z0-9]{6}$/.test(code)) {
    await sendMessage(chatId, "Отправь 6-значный код с экрана входа.");
    return;
  }

  try {
    const headers = { "Content-Type": "application/json" };
    if (BOT_SECRET) {
      headers["Authorization"] = `Bearer ${BOT_SECRET}`;
    }

    const res = await fetch(`${APP_URL}/api/auth/telegram/callback`, {
      method: "POST",
      headers,
      body: JSON.stringify({
        code,
        telegram_user: user.first_name + (user.last_name ? ` ${user.last_name}` : ""),
        telegram_id: user.id,
      }),
    });

    const data = await res.json();

    if (data.success) {
      await sendMessage(chatId, "Вход выполнен. Можешь вернуться в браузер.");
    } else {
      await sendMessage(chatId, "Код неверный или истёк. Попробуй ещё раз.");
    }
  } catch (err) {
    console.error("Callback error:", err.message);
    await sendMessage(chatId, "Ошибка связи с сервером. Попробуй позже.");
  }
}

async function poll() {
  while (true) {
    try {
      const res = await fetch(`${API}/getUpdates?offset=${offset}&timeout=30`);
      const data = await res.json();

      if (data.ok && data.result.length > 0) {
        for (const update of data.result) {
          offset = update.update_id + 1;
          if (update.message) {
            await handleMessage(update.message);
          }
        }
      }
    } catch (err) {
      console.error("Poll error:", err.message);
      await new Promise((r) => setTimeout(r, 5000));
    }
  }
}

console.log("[TelegramBot] Starting...");
console.log(`[TelegramBot] App URL: ${APP_URL}`);
poll();
